crates/pathfinder/src/rpc/serde.rs

```diff
@@ -196,6 +197,49 @@ impl<'de> Visitor<'de> for FeeVisitor {
     }
 }
 
+pub struct StarknetBlockNumberAsHexStr;
+
+impl SerializeAs<StarknetBlockNumber> for StarknetBlockNumberAsHexStr {
+    fn serialize_as<S>(source: &StarknetBlockNumber, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let bytes = source.0.to_be_bytes();
+        // StarknetBlockNumber is "0x" + 16 digits at most
+        let mut buf = [0u8; 2 + 16];
+        let s = bytes_as_hex_str(&bytes, &mut buf).map_err(serde::ser::Error::custom)?;
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de> DeserializeAs<'de, StarknetBlockNumber> for StarknetBlockNumberAsHexStr {
+    fn deserialize_as<D>(deserializer: D) -> Result<StarknetBlockNumber, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(StarknetBlockNumberVisitor)
+    }
+}
+
+struct StarknetBlockNumberVisitor;
+
+impl<'de> Visitor<'de> for StarknetBlockNumberVisitor {
+    type Value = StarknetBlockNumber;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("a hex string of up to 16 digits with an optional '0x' prefix")
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        u64::from_str_radix(v, 16)
+            .map_err(serde::de::Error::custom)
+            .map(StarknetBlockNumber)
```

Comment:
Trying to parse `"0x1234"` with `u64::from_str_radix(_, 16)` fails with invalid digit. Perhaps the expecting is wrong? Didn't yet see where this is used.